dice-game-ui-angular/src/app/game/game.component.ts

```diff
@@ -1,13 +1,334 @@
+          }
+        } else {
+          alert(`Selected area is not equal to ${this.dice1} by ${this.dice2}`);
+          this.clearSelection();
+        }
+
+        this.startPoint = null;
+      })
+      .on('mouseover', (d, i, nodes) => {
+        const rect = d3.select(nodes[i]);
+        this.select_area(rect);
+      })
+      .style('fill', this.emptyColor);
+
+    // update
+    this.svg
+      .selectAll('rect')
+      .data(this.points)
+      .style('fill', (d) => {
+        if (d.set === true) {
+          return this.setColor;
+        } else if (d.selected === true) {
+          return this.selectedColor;
+        } else {
+          return this.emptyColor;
+        }
+      });
+
+    // delete
+    this.svg.selectAll('rect').data(this.points).exit().remove();
+  }
+
+  public clearSelection() {
+    for (const point of this.points) {
+      point.selected = false;
+    }
+    this.render_field();
+  }
+
+  public isFullRect(rect: Rect): boolean {
+    const selectedPointsCount = this.points.filter((x) => x.selected).length;
+    return selectedPointsCount === this.dice1 * this.dice2;
+  }
+
+  public select_area(pointElement): Rect {
+    if (this.selecting) {
+      pointElement.style('fill', this.selectedColor);
+      return this.set_neighbors(this.startPoint, {
+        x: parseInt(pointElement.attr('data-point-x'), 10),
+        y: parseInt(pointElement.attr('data-point-y'), 10),
+        selected: false,
+        set: false,
+      });
+    }
+    return null;
+  }
+
+  public set_neighbors(startPoint: FieldPoint, endPoint: FieldPoint): Rect {
+    if (startPoint == null) {
+      return;
+    }
+    let d1 = this.dice1;
+    let d2 = this.dice2;
+    let minX = Math.min(startPoint.x, endPoint.x);
+    let minY = Math.min(startPoint.y, endPoint.y);
+    let maxX = Math.max(startPoint.x, endPoint.x);
+    let maxY = Math.max(startPoint.y, endPoint.y);
+    const lenX = Math.abs(minX - maxX) + 1;
+    const lenY = Math.abs(minY - maxY) + 1;
+
+    // Change dices in accordance with available length
+    if (Math.abs(d2 - lenX) < Math.abs(d1 - lenX)) {
+      const curD = d2;
+      d2 = d1;
+      d1 = curD;
+    }
+    // Limit width by dice value
+    if (lenX >= d1) {
+      if (minX === startPoint.x) {
+        maxX -= lenX - d1;
+      } else {
+        minX += lenX - d1;
+      }
+    }
+    // Limit height by dice value
+    if (lenY >= d2) {
+      if (minY === startPoint.y) {
+        maxY -= lenY - d2;
+      } else {
+        minY += lenY - d2;
+      }
+    }
+    // Set selection for points
+    for (const point of this.points) {
+      if (point.x <= maxX && point.x >= minX && point.y <= maxY && point.y >= minY) {
+        point.selected = true;
+      } else {
+        point.selected = false;
+      }
+    }
+
+    this.render_field();
+    return new Rect(minY, minX, maxY - minY + 1, maxX - minX + 1); // inverted axis!!
+  }
+}
 
-  ngOnInit(): void {}
+interface FieldPoint {
```

Comment:
It is good practice (I don't know whether it is common practice, but I always follow it) to define interface at the beginning of the file.

Reply:
ok. Fixed